src/full/Agda/Syntax/Parser/LexActions.hs

```diff
@@ -178,10 +181,31 @@ endWith a = popLexState `andThen` a
 
 -- | Parse a 'Keyword' token, triggers layout for 'layoutKeywords'.
 keyword :: Keyword -> LexAction Token
-keyword k = layout $ withInterval_ (TokKeyword k)
+keyword k =
+    case k of
+
+        -- Unconditional layout keyword.
+        _ | k `elem` layoutKeywords ->
+            withLayout k cont
+
+        -- @constructor@ is not a layout keyword in @record ... where@ blocks,
+        -- only in @interleaved mutual@ blocks.
+        KwConstructor -> do
+            cxt <- getContext
+            if inMutualAndNotInWhereBlock cxt
+              then withLayout k cont
+              else cont
+
+        _ -> cont
     where
-        layout | k `elem` layoutKeywords  = withLayout
-               | otherwise              = id
+    cont = withInterval_ (TokKeyword k)
+
+    -- Most recent block decides ...
+    inMutualAndNotInWhereBlock = \case
+      Layout KwMutual _ _ : _ -> True
+      Layout KwWhere  _ _ : _ -> False
+      _ : bs                  -> inMutualAndNotInWhereBlock bs
+      []                      -> False
```

Comment:
Should this last case be `True` so that we preserve good error messages in cases such
as Issue2858-invalid.agda?

Reply:
Ah, good point (although this does not help anymore if the user writes their top-level `module Top where`).  So maybe the practical improvement is minor. 
Of course, only keeping track of the layout keyword `where` is rather cheap and imprecise, (barely) sufficient to keep `record...where` and `interleaved mutual` apart.  
If we wanted better error messages, we could try to keep track of more context.  However, this is hard to do in the lexer, and having not to worry about such things inside `Parser.y` is nice for the developer.

Reply:
Pushed a patch with your suggestion.